### PR TITLE
[SPARK-33101][ML] Make LibSVM format propagate Hadoop config from DS options to underlying HDFS file system

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Propagate LibSVM options to Hadoop configs in the LibSVM datasource.

### Why are the changes needed?
There is a bug that when running:
```scala
spark.read.format("libsvm").options(conf).load(path)
```
The underlying file system will not receive the `conf` options.

### Does this PR introduce _any_ user-facing change?
Yes. After the changes, for example, users should read files from Azure Data Lake successfully:
```scala
def hadoopConf1() = Map[String, String](
  s"fs.adl.oauth2.access.token.provider.type" -> "ClientCredential",
  s"fs.adl.oauth2.client.id" -> dbutils.secrets.get(scope = "...", key = "..."),
  s"fs.adl.oauth2.credential" -> dbutils.secrets.get(scope = "...", key = "..."),
  s"fs.adl.oauth2.refresh.url" -> s"https://login.microsoftonline.com/.../oauth2/token")
val df = spark.read.format("libsvm").options(hadoopConf1).load("adl://....azuredatalakestore.net/foldersp1/...")
```
and not get the following exception because the settings above are not propagated to the filesystem:
```java
java.lang.IllegalArgumentException: No value for fs.adl.oauth2.access.token.provider found in conf file.
	at ....adl.AdlFileSystem.getNonEmptyVal(AdlFileSystem.java:820)
	at ....adl.AdlFileSystem.getCustomAccessTokenProvider(AdlFileSystem.java:220)
	at ....adl.AdlFileSystem.getAccessTokenProvider(AdlFileSystem.java:257)
	at ....adl.AdlFileSystem.initialize(AdlFileSystem.java:164)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2669)
``` 

### How was this patch tested?
Added UT to `LibSVMRelationSuite`.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala [hunk 1]
```diff
@@ -105,13 +105,15 @@ object MLUtils extends Logging {
...
+      sparkSession: SparkSession,
+      paths: Seq[String],
+      options: Map[String, String]): RDD[(Double, Array[Int], Array[Double])] = {
     val lines = sparkSession.baseRelationToDataFrame(
       DataSource.apply(
         sparkSession,
         paths = paths,
         className = classOf[TextFileFormat].getName,
-        options = Map(DataSource.GLOB_PATHS_KEY -> "false")
+        options = options ++ Map(DataSource.GLOB_PATHS_KEY -> "false")
```
Comment (MaxGekk) [inline mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala:116]: This is similar to CSV/JSON datasource: https://github.com/apache/spark/blob/954cd9feaa1a3d4ad9a235811ae58e02a63e8386/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala#L158

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala [hunk 2]
```diff
@@ -27,12 +27,13 @@ import org.apache.spark.ml.attribute.AttributeGroup
...
 import org.apache.spark.mllib.util.MLlibTestSparkContext
-import org.apache.spark.sql.{Row, SaveMode}
+import org.apache.spark.sql.{FakeFileSystemRequiringDSOption, Row, SaveMode}
+import org.apache.spark.sql.catalyst.plans.SQLHelper
 import org.apache.spark.sql.types.{DoubleType, StructField, StructType}
 import org.apache.spark.util.Utils
 
 
-class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext {
+class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext with SQLHelper {
```
Comment (MaxGekk) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala:36]: Added `SQLHelper` for `withSQLConf`

**Comment by MaxGekk** [comment]
@HyukjinKwon @dongjoon-hyun @srowen I have found one more datasource where options are not propagated to hadoop configs. I hope this is the last one. Please, review this PR. It is similar to https://github.com/apache/spark/pull/29976

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @MaxGekk and @HyukjinKwon .
Merged to master.

**Comment by MaxGekk** [comment]
Here is a backport to 3.0: https://github.com/apache/spark/pull/29986

**Pull request closed without merging**